src/general_tally.cpp

```diff
@@ -0,0 +1,210 @@
+    check_ifany_filter = false;
+  }
+
+  StaticVector4 tally_dimensions_(position_shape_.begin(),
+                                  position_shape_.end());
+
+  // get the size of the enery_filter
+  if (energy_in_) {
+    size_t ne = energy_in_->size();
+    tally_dimensions_.insert(tally_dimensions_.begin(), ne);
+
+    check_ifany_filter = check_ifany_filter && false;
+  }
+
+  // if any of the filter is not present, then give fatal_error
+  if (check_ifany_filter == true) {
+    fatal_error("for the " + name_ + " tally, no filter is provided.");
+  }
+
+  tally_avg.reallocate(tally_dimensions_);
+  tally_avg.fill(0.0);
+
+  tally_gen_score.reallocate(tally_dimensions_);
+  tally_gen_score.fill(0.0);
+
+  tally_var.reallocate(tally_dimensions_);
+  tally_var.fill(0.0);
+}
+
+void GeneralTally::score_collision(const Particle& p, const Tracker& tktr,
+                                   MaterialHelper& mat) {
+  if (!(estimator_ == GeneralTally::Estimator::Collision)) {
+    return;
+  }
+
+  // get the indexes for the positions, if exist
+  StaticVector3 position_indexes;
+  if (position_filter_) {
+    position_indexes = position_filter_->get_indices(
+        tktr);  // it will provde the reduce dimensions
+  }
+  StaticVector4 indexes(position_indexes.begin(), position_indexes.end());
+
+  // get the index for the energy if exist
+  std::size_t index_E;
+  if (energy_in_) {
+    std::optional<std::size_t> E_indx = energy_in_->get_index(p.E());
+    if (E_indx.has_value()) {
+      index_E = E_indx.value();
+    } else {
+      return;
+    }
+  }
+
+  if (indexes.empty()) {
+    return;
+  }
+
+  const double Et = mat.Et(p.E());
+  const double collision_score = particle_base_score(p, mat) / Et;
+
+#ifdef ABEILLE_USE_OMP
+#pragma omp atomic
+#endif
+  tally_gen_score(indexes) += collision_score;
+}
+
+void GeneralTally::score_flight(const Particle& p, const Tracker& trkr,
+                                double d_flight, MaterialHelper& mat) {
+  if (!(estimator_ == GeneralTally::Estimator::TrackLength)) {
+    return;
+  }
+
+  // get the energy index
+  std::size_t index_E;
+  if (energy_in_) {
+    // get the energy_index if we in the bounds
+    std::optional<std::size_t> E_indx = energy_in_->get_index(p.E());
+    if (E_indx.has_value()) {
+      index_E = E_indx.value();
+    } else {
+      return;
+    }
+  }
+
+  double flight_score = particle_base_score(p, mat);
+
+  // get the all the bins' index along with the distance travelled by the
+  // particle
+  std::vector<TracklengthDistance> pos_indexes_ =
+      position_filter_->get_indices_tracklength(trkr, d_flight);
+
+  for (size_t iter = 0; iter < pos_indexes_.size(); iter++) {
+    StaticVector4 all_indexes_(pos_indexes_[iter].indexes_.begin(),
+                               pos_indexes_[iter].indexes_.end());
+    const double d_ = pos_indexes_[iter].distance_in_bin;
+
+    if (energy_in_) {
+      all_indexes_.insert(all_indexes_.begin(), index_E);
+    }
```

Comment:
You should do this at the beginning of the for loop, instead of here, as this will eliminate the need to move all the other indices that you previously added.